sdk/python/kfp/v2/compiler/compiler.py

```
@@ -0,0 +1,843 @@
+      tuples (param_name, producing_op_name). producing_op_name is the name of the op that
+      produces the param. If the param is a pipeline param (no producer op), then
+      producing_op_name is None.
+    """
+    inputs = defaultdict(set)
+    outputs = defaultdict(set)
+
+    for op in pipeline.ops.values():
+      # op's inputs and all params used in conditions for that op are both considered.
+      from kfp.components.structures import ComponentSpec
+      for param in op.inputs + list(condition_params[op.name]):
+        # if the value is already provided (immediate value), then no need to expose
+        # it as input for its parent groups.
+        if param.value:
+          continue
+        if param.op_name:
+          upstream_op = pipeline.ops[param.op_name]
+          upstream_groups, downstream_groups = \
+            self._get_uncommon_ancestors(op_groups, opsgroup_groups, upstream_op, op)
+          for i, group_name in enumerate(downstream_groups):
+            if i == 0:
+              # If it is the first uncommon downstream group, then the input comes from
+              # the first uncommon upstream group.
+              inputs[group_name].add((param.full_name, upstream_groups[0]))
+            else:
+              # If not the first downstream group, then the input is passed down from
+              # its ancestor groups so the upstream group is None.
+              inputs[group_name].add((param.full_name, None))
+          for i, group_name in enumerate(upstream_groups):
+            if i == len(upstream_groups) - 1:
+              # If last upstream group, it is an operator and output comes from container.
+              outputs[group_name].add((param.full_name, None))
+            else:
+              # If not last upstream group, output value comes from one of its child.
+              outputs[group_name].add((param.full_name, upstream_groups[i+1]))
+        else:
+          if not op.is_exit_handler:
```

I think we can raise an error when `op.is_exit_handler` because it's yet to be supported.